structs/schedule.go

```diff
@@ -0,0 +1,248 @@
+func (s *SustainerItem) GetPageURL() string {
+	return ""
+}
+
+/*
+ * Implementation of ScheduleItem for TimeslotItem
+ */
+
+// GetName gets the display name of a TimeslotItem.
+func (t *TimeslotItem) GetName(context *PageContext) string {
+	return t.Timeslot.Title
+}
+
+// GetDesc gets the description of a TimeslotItem.
+func (t *TimeslotItem) GetDesc(context *PageContext) string {
+	return t.Timeslot.Description
+}
+
+// GetStart gets the start time of a TimeslotItem.
+func (t *TimeslotItem) GetStart() time.Time {
+	return t.Timeslot.StartTime
+}
+
+// GetFinish gets the finish time of a TimeslotItem.
+func (t *TimeslotItem) GetFinish() time.Time {
+	return t.Timeslot.StartTime.Add(t.Timeslot.Duration)
+}
+
+// GetBlock gets the block name of a TimeslotItem.
+func (t *TimeslotItem) GetBlock() string {
+	// TODO(CaptainHayashi): calculate schedule block here.
+	return "normal"
+}
+
+// IsSustainer gets whether a TimeslotItem is sustainer (it isn't).
+func (t *TimeslotItem) IsSustainer() bool {
+	return false
+}
+
+// HasPage gets whether a TimeslotItem has a show page (it does).
+func (t *TimeslotItem) HasPage() bool {
+	return true
+}
+
+// GetPageURL gets the root-relative URL to a TimeslotItem's show page.
+func (t *TimeslotItem) GetPageURL() string {
+	return t.PageURL
+}
+
+/*
+ * Schedule filling
+ * TODO(CaptainHayashi): This DEFINITELY doesn't belong in structs
+ */
+
+// FillTimeslotSlice converts a slice of Timeslots to a slice of ScheduleItems.
+// It does so by filling in any gaps between the start time and the first show, the final show and the finish time, and any two shows.
+// It expects a constructor function for lifting Timeslots to TimeslotItems.
+// It will return an error if any two shows overlap.
+// It presumes the timeslot slice is already sorted in chronological order.
+func FillTimeslotSlice(start, finish time.Time, slots []myradio.Timeslot, tbuilder func(*myradio.Timeslot) (*TimeslotItem, error)) ([]ScheduleItem, error) {
+	nslots := len(slots)
+
+	// The maximum possible number of items is 2(nslots) + 1:
+	// nslots slots, (nslots - 1) sustainers in between, and 2 sustainers at the ends.
+	items := make([]ScheduleItem, (2*len(slots))+1)
+
+	// Now deal with the easy case--no slots.
+	if nslots == 0 {
+		items[0] = NewSustainerItem(start, finish)
+		return items, nil
+	}
+
+	// Otherwise, we now have to do some actual filling.
+	i := 0
+
+	// First, work out if we need to fill before the first show.
+	firstShow := slots[0]
+	if start.Before(firstShow.StartTime) {
+		items[i] = NewSustainerItem(start, firstShow.StartTime)
+		i++
+	}
+
+	// Now, if possible, start filling between.
+	// This will add all but the last show.
+	var err error
+	for j := range slots {
+		if j < nslots-1 {
+			first := &slots[j]
+			second := &slots[j+1]
+
+			firstFinish := first.StartTime.Add(first.Duration)
+			if firstFinish.After(second.StartTime) {
+				return nil, fmt.Errorf(
```

Comment:
We should probably recover more gracefully from this, as it happens a lot in real life